### PR TITLE
[WIP] Process free attestation

### PR DESCRIPTION
**Pull request by g-r-a-n-t**
## Issue Addressed
#406 

## Proposed Changes
Adds an attestation to `fork_choice` whenever it's processed by the beacon chain.

## Additional Info
Took the liberty of renaming `ForkChoice::process_attestation_from_block` to `process_attestation` and made it public, wasn't sure if there's a reason for not doing this. 

It's checking the signature twice here, once on `beacon_chain.rs:497` and again on line `503`. This seems like something that would be nice to avoid, but not sure of a good way of doing so.

Didn't expand the API in `ForkChoice` to determine whether or not the message is latest, will do this after the rest is ok.

### CODE REVIEW

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 1]
```diff
@@ -480,6 +480,24 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         self.metrics.attestation_processing_requests.inc();
         let timer = self.metrics.attestation_processing_times.start_timer();
 
+        // Retrieve the attestation's state from `store` if necessary.
+        let attestation_state = match attestation.data.beacon_block_root == self.canonical_head.read().beacon_block_root {
```
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: This works, but we could loosen this `if` statement and do less reads from the database.

A `BeaconState` can be used to verify an `Attestation` if both of these conditions are true:

- The `BeaconState` and `Attestation` are "age appropriate". I.e., `epoch(attestation_slot)` is either equal or adjacent to `epoch(state.slot)`.
- The `BeaconState` block history contains the `target.root`. (You can use [`BestBlockRootsIterator`](https://github.com/sigp/lighthouse/blob/051355925299862ea2298c6f5d24fdf1dfc6fbd0/beacon_node/store/src/iter.rs#L165-L188) to do this quite easily).

If either of those two conditions are not met, then we use `target.root` to load the `BeaconBlock`, then use `block.state_root` to load a `BeaconState` that is necessarily a valid state to use. (Note: I think using `attestation.data.target.root` is better than `attestation.data.beacon_block_root`, as I had previously said).

**Comment by g-r-a-n-t** [comment]
This is good for another review :) Updating the target will clean this up.

**Comment by paulhauner** [comment]
Hey, sorry I missed your previous comment!

**Comment by paulhauner** [comment]
Hey, sorry @g-r-a-n-t can you please merge/rebase with `master`. The diff is a bit noisy at the moment :)

**Comment by g-r-a-n-t** [comment]
Hey @paulhauner, I didn't realize the branch with the fork choice tests got merged, my bad.

Also, I've got a bit of a mess on my hands since I had merged in half of the changes from `fork-choice-tests` before making changes to my branch and then just now merging master, which had the changes from `fork-choice-tests` compressed into a single commit. Reconciling this may be a little involved and I may need to create a new PR :P I'll try to get this figured out tomorrow.

**Comment by paulhauner** [comment]
Bummer, sorry to hear about that!

Let me know if I can help :)

**Comment by g-r-a-n-t** [comment]
Hey @paulhauner, this is all cleaned up and ready to go!

File: eth2/lmd_ghost/src/reduced_tree.rs [hunk 2]
```diff
@@ -109,6 +109,10 @@ where
             .update_root(new_block.slot, new_root)
             .map_err(|e| format!("update_finalized_root failed: {:?}", e))
     }
+
+    fn latest_message(&self, validator_index: usize) -> Option<(Hash256, Slot)> {
+        self.core.write().latest_message(validator_index)
```
[the remark this answers is not in the document]
Comment (g-r-a-n-t) [reply eth2/lmd_ghost/src/reduced_tree.rs]: Yeah, I remember the elastic array's dynamic sizing throwing a wrench in this.

**Pull request closed without merging**